### PR TITLE
[MTS/TL/HH][111418676] Fetch data using data attrs

### PR DESCRIPTION
**Pull request by michaelstalker**
https://www.pivotaltracker.com/story/show/111418676

**This PR looks very large. The code changes to the source files only constitute a small portion of this PR, however. Most changes are to the compiled code in the `/dist` folder, or to the tests.**

We need to find an element's item, section (which we call the subgroup
inside the jquery-autotagging.js codebase), and value. This code change
is based off of Rent.com's JavaScript tracking code.

We may need to implement a more nuanced way of finding the clicked
element; the Rent.com code deals with the target's parent element when
deciding what the click target is.

### CODE REVIEW

File: src/data_with_data_attributes.coffee [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+], ($) ->
+
+  (itemDataAttribute, sectionDataAttribute) ->
+    isSelect = (node) -> node.nodeName is 'SELECT'
+
+    isInput = (node) ->
+      isSelect(node) || node.nodeName is 'INPUT' || node.nodeName is 'TEXTAREA'
+
+    value: (node) ->
+      if isInput(node)
```
Comment (warrenv) [inline src/data_with_data_attributes.coffee:12]: What about using 'text = if' here. Then 'text' could be removed in each of the condition blocks.

File: src/jquery.autotagging.coffee [hunk 2]
```diff
@@ -24,22 +36,24 @@ define [
...
-      @fireCallback      = opts.fireCallback
-      @path              = "#{document.location.pathname}#{document.location.search}"
-      @warehouseURL      = opts.warehouseURL
-      @opts              = opts
+      @exclusionList       = opts.exclusionList || []
+      @fireCallback        = opts.fireCallback
+      @path                = "#{document.location.pathname}#{document.location.search}"
+      @warehouseURL        = opts.warehouseURL
       @followHref = if opts.followHref? then opts.followHref else true
+      useDataTags          = opts.useDataTags || false # opts.itemDataAttribute && opts.sectionDataAttribute
```
Comment (warrenv) [inline src/jquery.autotagging.coffee]: Is the comment still needed?

File: src/select_data_without_data_attributes.coffee [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  textFormatter
+) ->
+  subgroup: ($elem) ->
+    $elem.closest('[id]').attr('id') or ''
+
+  item: ($elem) ->
+    value = $elem.find(':selected').val() or ''
+    textFormatter.replaceDoubleByteChars(value)
+
+  value: ($elem) ->
```
Comment (warrenv) [inline src/select_data_without_data_attributes.coffee:14]: Can this and the code starting at line 11 be refactored? It looks like they both do the same thing.

**Comment by warrenv** [comment]
There are several tests that seem like duplicates on first look. Not sure if we want to clean those up or not. For example:

``` bash
spec/javascripts/click_data_without_data_attributes_spec.js
46:    it('returns an empty string if there is no subgroup', function () {

spec/javascripts/data_with_data_attributes_spec.js
35:    it('returns an empty string if there is no subgroup', function () {

spec/javascripts/select_data_without_data_attributes_spec.js
77:    it('returns an empty string if there is no subgroup', function () {

```

File: src/jquery.autotagging.coffee (hunk 2 above)
Comment (michaelstalker) [inline src/jquery.autotagging.coffee]: Nope! I'll remove it.

File: src/data_with_data_attributes.coffee (hunk 1 above)
Comment (michaelstalker) [inline src/data_with_data_attributes.coffee:12]: :+1:

File: src/select_data_without_data_attributes.coffee (hunk 3 above)
Comment (michaelstalker) [inline src/select_data_without_data_attributes.coffee:14]: :+1:

File: src/data_with_data_attributes.coffee (hunk 1 above)
Comment (michaelstalker) [inline src/data_with_data_attributes.coffee:12]: I'll replace the temporary variable with a call to a [query function](http://refactoring.com/catalog/replaceTempWithQuery.html).

**Comment by michaelstalker** [comment]
Thanks for reviewing this! I updated the code. My preference would be to maintain the duplicated tests for an empty string value. The functions in the three files return the same value, but only because I think, there's a business rule for each saying this should happen. The default value for one could change independently of the others. What are your thoughts?

**Comment by warrenv** [comment]
:+1:

**Comment by HansHauge** [comment]
:+1:

**Comment by Teapane** [comment]
:+1:

**Pull request merged**